### PR TITLE
7513 oauth backend quickbooks

### LINKED ISSUE #7513: oAuth backend for quickbooks source
Part of #7000 story

Implement quickbooks oauth backend and do e2e test.

### PR DESCRIPTION
**Pull request by eliziario**
## What
Implements OAuth Backend on server for Quickboooks Source. 
https://github.com/airbytehq/airbyte/issues/7513

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java [hunk 1]
```diff
@@ -8,63 +8,65 @@
...
+    return create(sourceDefinition.getDockerRepository());
+  }
+
+  public OAuthFlowImplementation create(final StandardDestinationDefinition destinationDefinition) {
+    return create(destinationDefinition.getDockerRepository());
   }
 
-  public OAuthFlowImplementation create(final StandardDestinationDefinition destinationDefinition, final UUID workspaceId) {
-    final String imageName = destinationDefinition.getDockerRepository();
+  public OAuthFlowImplementation create(final String imageName) {
```
Comment (avida) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java:61]: please make it private

File: airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java [hunk 2]
```diff
@@ -8,63 +8,65 @@
...
-        .put("airbyte/source-intercom", new IntercomOAuthFlow(configRepository, httpClient))
-        .put("airbyte/source-instagram", new InstagramOAuthFlow(configRepository, httpClient))
-        .put("airbyte/source-salesforce", new SalesforceOAuthFlow(configRepository, httpClient))
-        .put("airbyte/source-slack", new SlackOAuthFlow(configRepository, httpClient))
-        .put("airbyte/source-surveymonkey", new SurveymonkeyOAuthFlow(configRepository, httpClient))
-        .put("airbyte/source-trello", new TrelloOAuthFlow(configRepository, httpClient))
+    this.configRepository = configRepository;
+    this.httpClient = httpClient;
+    OAUTH_FLOW_MAPPING = ImmutableMap.<String, Class<? extends OAuthFlowImplementation>>builder()
+        .put("airbyte/source-asana", AsanaOAuthFlow.class)
```
Comment (avida) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java:33]: so you want to create new instance on every call? It  It would break trello cause it using oAuth1.0 and storing some parameters from previos step.

**Review by avida — CHANGES_REQUESTED** [review]
Making new instance on each call would break trello oAuth1.0 flow.

File: airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java (hunk 2 above)
Comment (avida) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java:33]: Looks like I was wrong, it wont break trello cause it already using static variables for saving state. Can we make OAUTH_FLOW_MAPPING static final ?

**Review by avida — APPROVED** [review]

File: airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java [hunk 3]
```diff
@@ -8,65 +8,65 @@
...
 import java.net.http.HttpClient;
 import java.util.Map;
-import java.util.UUID;
 
 public class OAuthImplementationFactory {
 
-  private final Map<String, OAuthFlowImplementation> OAUTH_FLOW_MAPPING;
+  private static final Map<String, Class<? extends OAuthFlowImplementation>> OAUTH_FLOW_MAPPING =
+      ImmutableMap.<String, Class<? extends OAuthFlowImplementation>>builder()
+          .put("airbyte/source-asana", AsanaOAuthFlow.class)
```
Comment (sherifnada) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java:27]: what is the rationale behind this change? I think I prefer it less than the previous approach because it uses reflection, which means tracking how this code is being used is harder to do in the IDE. Also, reflection can fail at runtime, whereas the constructor usage is checked at compile time

**Review by sherifnada — COMMENTED** [review]
question about the change to the factory but the oauth impl itself lgtm

File: airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java (hunk 3 above)
Comment (eliziario) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java:27]: Yes, no very strong reasons. Personally I think this is more readable, but it is really personal preference. And your concerns are a good point, let me change it back to the way it was.

**Review by sherifnada — APPROVED** [review]
